src/basic/confidential-virt.c

```diff
@@ -194,40 +195,62 @@ static bool detect_hypervisor(void) {
         return is_hv;
 }
 
-ConfidentialVirtualization detect_confidential_virtualization(void) {
-        static thread_local ConfidentialVirtualization cached_found = _CONFIDENTIAL_VIRTUALIZATION_INVALID;
+static ConfidentialVirtualization detect_confidential_virtualization_impl(void) {
         char sig[13] = {};
-        ConfidentialVirtualization cv = CONFIDENTIAL_VIRTUALIZATION_NONE;
-
-        if (cached_found >= 0)
-                return cached_found;
 
         /* Skip everything on bare metal */
         if (detect_hypervisor()) {
                 cpuid_leaf(0, sig, true);
 
                 if (memcmp(sig, CPUID_SIG_AMD, sizeof(sig)) == 0)
-                        cv = detect_sev();
+                        return detect_sev();
                 else if (memcmp(sig, CPUID_SIG_INTEL, sizeof(sig)) == 0)
-                        cv = detect_tdx();
+                        return detect_tdx();
         }
 
-        cached_found = cv;
-        return cv;
+        return CONFIDENTIAL_VIRTUALIZATION_NONE;
+}
+#elif defined(__s390x__)
+static ConfidentialVirtualization detect_confidential_virtualization_impl(void) {
+        _cleanup_free_ char *s = NULL;
+        size_t readsize;
+        int r;
+
+        r = read_full_virtual_file("/sys/firmware/uv/prot_virt_guest", &s, &readsize);
+        if (r < 0) {
+                log_debug_errno(r, "Unable to read /sys/firmware/uv/prot_virt_guest: %m");
```

Comment:
I imagine this file is only present on certain s390x setups? If so, the log should be behind an `if (r != -ENOENT)` as there's no point in logging about it not being found

Reply:
The kernel made it a bit wierd in that the entire `/sys/firmware/uv` tree will be absent if no ultravisor is present, but when the ultravisor is present, this subtree will be present for both host and guest kernels.

* Old kernel -> file absent
* New kernel + no ultravisor -> subtree absent in both host and guest
* New kernel + ultravisor -> subtree present in both host and guest. 'prot_virt_host' and 'prot_virt_guest' files distinguish host vs guest context.

IOW, i think having the debug log even for ENOENT is desirable, as that indicates that we've checked for the file and found it absent, confirming a non-confidential guest.

Reply:
ok